C/common/filter_plugin.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ * FogLAMP plugin filter class
+ *
+ * Copyright (c) 2018 Dianomic Systems
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Massimiliano Pinto
+ */
+
+#include <filter_plugin.h>
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+
+using namespace std;
+
+// Constructor
+FilterPlugin::FilterPlugin(const std::string& name,
+			   PLUGIN_HANDLE handle) : Plugin(handle), m_name(name)
+{
+	// Setup the function pointers to the plugin
+	pluginInit = (PLUGIN_HANDLE (*)(const ConfigCategory *,
+					OUTPUT_HANDLE *,
+					OUTPUT_STREAM output))
+					manager->resolveSymbol(handle,
+							       "plugin_init");
+  	pluginShutdownPtr = (void (*)(PLUGIN_HANDLE))
+				      manager->resolveSymbol(handle,
+							     "plugin_shutdown");
+  	pluginIngestPtr = (void (*)(PLUGIN_HANDLE, READINGSET *))
+				      manager->resolveSymbol(handle,
+							     "plugin_ingest");
+}
+
+// Destructor
+FilterPlugin::~FilterPlugin()
+{
+}
+
+// Init
+PLUGIN_HANDLE FilterPlugin::init(const ConfigCategory& config,
+				 OUTPUT_HANDLE *outHandle,
+				 OUTPUT_STREAM outputFunc)
+{
+	m_instance = this->pluginInit(&config,
+				      outHandle,
+				      outputFunc);
+	return &m_instance;
+}
+
+// Shutdown
+void FilterPlugin::shutdown()
+{
+	return this->pluginShutdownPtr(m_instance);
+}
+
+// Ingest readings through filters chain
+void FilterPlugin::ingest(READINGSET* readings)
+{
+        return this->pluginIngestPtr(m_instance, readings);
+}
+
+/**
+ * Load the specified filter plugin
+ *
+ * Static method
+ *
+ * @param    filterName		The filter plugin to load
+ * @return   Plugin handle on success, NULL otherwise 
+ *
+ */
+PLUGIN_HANDLE FilterPlugin::loadFilterPlugin(const string& filterName)
+{
+	if (filterName.empty())
+	{
+		Logger::getLogger()->error("Unable to fetch filter plugin '%s' from configuration.",
+			filterName.c_str());
+		// Failure
+		return NULL;
+	}
+	Logger::getLogger()->info("Loading filter plugin '%s'.", filterName.c_str());
+
+	PluginManager* manager = PluginManager::getInstance();
+	PLUGIN_HANDLE handle;
+	if ((handle = manager->loadPlugin(filterName, PLUGIN_TYPE_FILTER)) != NULL)
+	{
+		// Suceess
+		Logger::getLogger()->info("Loaded filter plugin '%s'.", filterName.c_str());
+	}
+	return handle;
+}
+
+/**
+ * Cleanup all the load filters setup
+ *
+ * Call "plugin_shutdown" method and free the FilterPlugin object
+ *
+ * Static method
+ *
+ * @param loadedFilters		The vector of loaded filters
+ *
+ */
+void FilterPlugin::cleanupFilters(std::vector<FilterPlugin *>& loadedFilters)
+{
+	// Cleanup filters
+	for (auto it = loadedFilters.begin(); it != loadedFilters.end(); ++it)
+	{
+		// Call filter plugin shutdown
+		(*it)->shutdown();
+		// Free filter
+		delete *it;
+	}
+}
+
+/**
+ * Load all filter plugins found in the configuration category
+ *
+ * Static method
```

Comment:
Yes: it can be called by any service, just pass the filter vector reference.

Purpose it's not to duplicate this code in all south services or north ones

Same for loadFilters  and oadFilterPlugin

They can be called by any service